### PR TITLE
Batcher respect message count & size bytes

### PR DESCRIPTION
**Pull request by tock-ibm**
addresses both #341 and #327

### CODE REVIEW

File: internal/bft/sched_test.go [hunk 1]
```diff
@@ -15,6 +15,7 @@ import (
 )
 
 func TestSchedule(t *testing.T) {
+	t.Skip("flaky")
```
Comment (yacovm) [inline internal/bft/sched_test.go:18]: I've never seen this fail in CI before...

Comment (tock-ibm) [reply internal/bft/sched_test.go:18]: There is always a first time, sorry

Comment (tock-ibm) [reply internal/bft/sched_test.go:18]: This object is not in use, anyway

Comment (tock-ibm) [reply internal/bft/sched_test.go:18]: The flaky test will be fixed in #242 
It hangs in unit tests in the IDE

Comment (HagarMeir) [reply internal/bft/sched_test.go:18]: it doesn't for me...weird

Comment (tock-ibm) [reply internal/bft/sched_test.go:18]: When it is a timing issue, it can happen very rarely

File: internal/bft/requestpool.go [hunk 2]
```diff
@@ -148,20 +150,29 @@ func (rp *Pool) Size() int {
...
+	count := minInt(rp.fifo.Len(), maxCount)
+	var totalSize uint64
+	batch = make([][]byte, 0, count)
 	var element = rp.fifo.Front()
-	for i := 0; i < m; i++ {
-		buff[i] = append(make([]byte, 0), element.Value.(*requestItem).request...)
+	for i := 0; i < count; i++ {
+		req := element.Value.(*requestItem).request
+		reqLen := uint64(len(req))
+		rp.logger.Debugf("i=%d, len(r)=%d, total=%d", i, len(req), totalSize)
```
Comment (yacovm) [inline internal/bft/requestpool.go:166]: I'm not sure this is the ideal place for this logging. I think ideally we will have a logging just before each place we return from the function (2 places total)

File: internal/bft/batcher.go [hunk 3]
```diff
@@ -12,64 +12,54 @@ import (
...
 		case <-b.closeChan:
 			return nil
 		case <-timeout:
-			return b.buildBatch(remainderOccupied, currBatch)
+			currBatch, _ = b.pool.NextRequests(b.maxMsgCount, b.maxSizeBytes)
+			return currBatch
 		default:
-			if b.pool.Size() >= b.batchSize-remainderOccupied {
-				return b.buildBatch(remainderOccupied, currBatch)
+			time.Sleep(10 * time.Millisecond)
```
Comment (yacovm) [inline internal/bft/batcher.go:52]: i think long term we should restructure this whole thing so we won't have to sleep at all, but have some kind of synchronization that eliminates sleeping.

**Review by yacovm — APPROVED** [review]

File: internal/bft/requestpool.go (hunk 2 above)
Comment (tock-ibm) [reply internal/bft/requestpool.go:166]: Correct. I added this during debug and forgot to remove. I'll push an update.

File: internal/bft/batcher.go (hunk 3 above)
Comment (tock-ibm) [reply internal/bft/batcher.go:52]: I agree.

**Pull request merged**